File: packages/python/src/xyp/_signer.py
"""Request signing: the three HTTP headers XYP requires on every call."""

from __future__ import annotations

import os
import time
from base64 import b64encode
from pathlib import Path

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.primitives.asymmetric.rsa import RSAPrivateKey

from xyp.errors import XypConfigError

PrivateKeySource = str | bytes | os.PathLike[str]

_PEM_MARKER = "-----BEGIN"


class Signer:
    """Holds the access token and the parsed RSA key; produces fresh auth headers."""

    def __init__(
        self,
        access_token: str,
        private_key: PrivateKeySource,
        password: bytes | None = None,
    ) -> None:
        if not access_token:
            raise XypConfigError("access_token must not be empty")
        self._access_token = access_token
        self._key = _load_private_key(private_key, password)

    def __repr__(self) -> str:
        return "Signer(access_token=***, private_key=***)"

    def headers(self, timestamp: int | None = None) -> dict[str, str]:
        """Headers for one request. XYP rejects stale timestamps, so never reuse them."""
        stamp = str(int(time.time()) if timestamp is None else timestamp)
        message = f"{self._access_token}.{stamp}".encode()
        signature = self._key.sign(message, padding.PKCS1v15(), hashes.SHA256())
        return {
            "accessToken": self._access_token,
            "timeStamp": stamp,
            "signature": b64encode(signature).decode("ascii"),
        }


def _read_key_bytes(source: PrivateKeySource) -> bytes:
    if isinstance(source, bytes):
        return source
    if isinstance(source, str) and _PEM_MARKER in source:
        return source.encode("ascii")
    try:
        return Path(source).read_bytes()
    except OSError as error:
        raise XypConfigError(f"Cannot read private key file: {error.strerror}") from error


def _load_private_key(source: PrivateKeySource, password: bytes | None) -> RSAPrivateKey:
    data = _read_key_bytes(source)
    is_pem = _PEM_MARKER.encode("ascii") in data
    loader = serialization.load_pem_private_key if is_pem else serialization.load_der_private_key
    try:
        key = loader(data, password=password)
    except (ValueError, TypeError):
        # Deliberately no detail from `error`: it can echo key material.
        raise XypConfigError(
            "private_key is not a valid PEM/DER private key (or the password is wrong)"
        ) from None
    if not isinstance(key, RSAPrivateKey):
        raise XypConfigError("private_key must be an RSA key")
    return key

File: packages/go/xyp/signer.go
package xyp

import (
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/pem"
	"errors"
	"io/fs"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
)

// XYP requires these three headers on every call. Their names are written in
// mixed case on the wire, and Go's http package would canonicalise them to
// "Accesstoken" and friends, so they are assigned to the header map directly.
const (
	accessTokenHeader = "accessToken"
	timeStampHeader   = "timeStamp"
	signatureHeader   = "signature"
)

const (
	pemPrivateKey          = "PRIVATE KEY"
	pemRSAPrivateKey       = "RSA PRIVATE KEY"
	pemEncryptedPrivateKey = "ENCRYPTED PRIVATE KEY"
	// procTypeHeader marks a legacy OpenSSL key encrypted with a passphrase.
	procTypeHeader = "Proc-Type"
	pemMarker      = "-----BEGIN"
)

// LoadPrivateKey reads an RSA private key from a PEM or DER file.
func LoadPrivateKey(path string) (crypto.Signer, error) {
	data, err := os.ReadFile(path) //nolint:gosec // the path is the caller's own configuration
	if err != nil {
		// Only the failure class, never the path's contents or the key itself.
		return nil, configErrorf("cannot read the private key file (%s)", fileProblem(err))
	}
	return ParsePrivateKey(data)
}

// ParsePrivateKey reads an RSA private key from PEM text or DER bytes.
func ParsePrivateKey(data []byte) (crypto.Signer, error) {
	if strings.Contains(string(data), pemMarker) {
		return parsePEM(data)
	}
	// DER names no format, so both common RSA encodings are tried.
	return firstRSAKey(parsePKCS8(data), parsePKCS1(data))
}

func parsePEM(data []byte) (crypto.Signer, error) {
	block, _ := pem.Decode(data)
	if block == nil {
		return nil, notAPrivateKey()
	}
	if block.Type == pemEncryptedPrivateKey || block.Headers[procTypeHeader] != "" {
		return nil, configErrorf(
			"the private key is encrypted, which this SDK cannot decrypt. Decrypt it once with " +
				"`openssl pkey -in encrypted.key -out private.key`, or pass a crypto.Signer of " +
				"your own (an HSM or KMS key works the same way) as Options.PrivateKey",
		)
	}
	switch block.Type {
	case pemPrivateKey:
		return firstRSAKey(parsePKCS8(block.Bytes))
	case pemRSAPrivateKey:
		return firstRSAKey(parsePKCS1(block.Bytes))
	default:
		return firstRSAKey(parsePKCS8(block.Bytes), parsePKCS1(block.Bytes))
	}
}

func parsePKCS8(der []byte) crypto.PrivateKey {
	key, err := x509.ParsePKCS8PrivateKey(der)
	if err != nil {
		return nil
	}
	return key
}

func parsePKCS1(der []byte) crypto.PrivateKey {
	key, err := x509.ParsePKCS1PrivateKey(der)
	if err != nil {
		return nil
	}
	return key
}

// firstRSAKey takes the first attempt that parsed. The underlying parse errors
// are deliberately dropped: they can echo key material into a log.
func firstRSAKey(attempts ...crypto.PrivateKey) (crypto.Signer, error) {
	for _, key := range attempts {
		if key == nil {
			continue
		}
		rsaKey, isRSA := key.(*rsa.PrivateKey)
		if !isRSA {
			return nil, notAnRSAKey()
		}
		return rsaKey, nil
	}
	return nil, notAPrivateKey()
}

func notAPrivateKey() *ConfigError {
	return configErrorf("PrivateKey is not a valid PEM or DER private key")
}

func notAnRSAKey() *ConfigError {
	return configErrorf("PrivateKey must be an RSA key")
}

func fileProblem(err error) string {
	switch {
	case errors.Is(err, fs.ErrNotExist):
		return "no such file"
	case errors.Is(err, fs.ErrPermission):
		return "permission denied"
	default:
		return "unreadable"
	}
}

// signHeaders builds the three credentials headers for one request. XYP rejects
// stale timestamps, so they are never reused between requests.
func (c *Client) signHeaders(header http.Header) error {
	timestamp := strconv.FormatInt(c.now().Unix(), 10)
	digest := sha256.Sum256([]byte(c.accessToken + "." + timestamp))
	signature, err := c.key.Sign(rand.Reader, digest[:], crypto.SHA256)
	if err != nil {
		return configErrorf("the private key could not sign the request")
	}
	// Assigned directly, not through Header.Set, so Go keeps the mixed case XYP expects.
	header[accessTokenHeader] = []string{c.accessToken}
	header[timeStampHeader] = []string{timestamp}
	header[signatureHeader] = []string{base64.StdEncoding.EncodeToString(signature)}
	return nil
}

// defaultClock is injectable through Client.now so tests can pin the timestamp.
func defaultClock() time.Time { return time.Now() }
